Extension/src/Utility/Process/program.ts

```diff
@@ -0,0 +1,172 @@
+import { Process } from './process';
+import { ReadableLineStream } from './streams';
+
+type ArrayPlusOptions<T, TOptions> = [...T[]] | [...T[], TOptions];
+
+function primitives(args: unknown[] | undefined): Primitive[] {
+    return args?.length ? args.filter(is.primitive) as Primitive[] : [];
+}
+
+export const searchPaths = lazy(() => filterToFolders(pathsFromVariable('PATH')));
+
+interface ProgramOptions {
+    cwd?: string;
+    env?: NodeJS.ProcessEnv;
+    noninteractive?: boolean;
+    on?: ArbitraryObject;
+    choices?: Promise<Set<string>>;
+}
+
+function options<T extends Record<string, any>>(args?: [...Primitive[]] | [...Primitive[], T]): Partial<T> {
+    return args?.length ? is.primitive(args[args.length - 1]) ? {} : args.pop() as T : {};
+}
+
+interface Launcher {
+    executable: Promise<string>;
+    cmdlineArgs: Primitive[];
+    options: ProgramOptions;
+}
+
+interface CommandResult {
+    code: number;
+    stdio: ReadableLineStream;
+    error: ReadableLineStream;
+}
+
+/**
+ * A function that when called will execute a program, and allow the consumer to interact with it.
+ *
+ * @param args the command line arguments to pass to the program.
+ * @returns a promise that resolves to a Process object, which can be used to interact with the program.
+ */
+export interface ProcessFunction extends Launcher {
+    (...commandLineArguments: ArrayPlusOptions<Primitive, ProgramOptions>): Promise<Process>;
+}
+
+/**
+ * A function that when called will execute a program, and return the CommandResult
+ *
+ * @param args the command line arguments to pass to the program.
+ * @returns a promise that resolves to a CommandResult object, which can be used to get the results of the program.
+ */
+export interface CommandFunction extends Launcher {
+    (...commandLineArguments: ArrayPlusOptions<Primitive, ProgramOptions>): Promise<CommandResult>;
+}
+
+async function processFactory(executable: string | Launcher, ...initialArgs: [...Primitive[], ProgramOptions & { noninteractive: true }]): Promise<CommandFunction>;
+async function processFactory(executable: string | Launcher, ...initialArgs: ArrayPlusOptions<Primitive, ProgramOptions>): Promise<ProcessFunction>;
+async function processFactory(executable: string | Launcher, ...initialArgs: ArrayPlusOptions<Primitive, ProgramOptions>): Promise<ProcessFunction | CommandFunction> {
+    let cmdlineArgs = primitives(initialArgs);
+    let opts = options(initialArgs);
+    let fullPath: Promise<string>;
+
+    if (typeof (executable) === 'string') {
+        fullPath = lazy<string>(async () => {
+            if (!await filepath.isExecutable(executable)) {
+                // if they didn't pass in a valid executable path, let's see if we can figure it out.
+
+                // if we were handed some choices, we'll look at them, otherwise we'll see what we can find on the PATH.
+                opts.choices ??= lazy(async () => new Finder(executable).scan(... await searchPaths).results);

```

Comment:
Normally, there is no space after `...`

Reply:
I've added yet another eslint rule for that.